File: processor/src/main/java/com/satis/viewmodel/processor/ObserveStoreCreatorProxy.java
package com.satis.viewmodel.processor;

import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.ParameterizedTypeName;
import com.squareup.javapoet.TypeSpec;
import com.squareup.javapoet.TypeVariableName;

import java.lang.reflect.Constructor;
import java.util.LinkedHashMap;

import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;

public class ObserveStoreCreatorProxy {
    private TypeSpec.Builder builder;
    private MethodSpec.Builder storeMethod;

    public ObserveStoreCreatorProxy() {
        initDefaultTypeBuilder();
    }

    private void initDefaultTypeBuilder(){
        builder = TypeSpec.classBuilder("ObserveStore");
        ClassName observeStore = ClassName.get("com.satis.viewmodel.core", "ObserveStore");
        builder.addSuperinterface(observeStore);
        builder.addModifiers(Modifier.PUBLIC);
        overrideMethod();
    }

    private void overrideMethod(){
        ClassName linkedHashMapTypeName = ClassName.get(LinkedHashMap.class);
        ClassName classTypeName = ClassName.get(Class.class);
        TypeVariableName lifeOwner = TypeVariableName.get("? extends androidx.lifecycle.LifecycleOwner");
        ParameterizedTypeName classParameterizedType = ParameterizedTypeName.get(classTypeName,lifeOwner.box());

        ClassName constructorTypeName = ClassName.get(Constructor.class);
        TypeVariableName observerTypeName = TypeVariableName.get("? extends com.satis.viewmodel.core.Observer");
        ParameterizedTypeName constructorParameterizedType =  ParameterizedTypeName.get(constructorTypeName,observerTypeName.box());
        ParameterizedTypeName linkedHashMapType =  ParameterizedTypeName.get(linkedHashMapTypeName, classParameterizedType,constructorParameterizedType);
        storeMethod = MethodSpec.methodBuilder("getStore")
                .addAnnotation(Override.class)
                .addModifiers(Modifier.PUBLIC)
                .returns(linkedHashMapType)
                .addStatement("$T map = new $T()",linkedHashMapType,linkedHashMapType)
                .beginControlFlow("try");
    }

    public void put(TypeElement typeElement){
        ClassName host = ClassName.bestGuess(typeElement.getQualifiedName().toString());
        ClassName className = ClassName.get(Processor.APT_PACKAGE, typeElement.getSimpleName().toString() + "$Observe");
        storeMethod.addStatement("map.put($T.class,$T.class.getConstructor())",host,className);
    }

    public void endMethod(){
        storeMethod.endControlFlow()
                .beginControlFlow("catch($T e)",Exception.class)
                .endControlFlow()
                .addStatement("return map");
    }

    public TypeSpec build(){
        builder.addMethod(storeMethod.build());
        return builder.build();
    }
}

File: viewmodel-processor/src/main/java/com/satis/viewmodel/processor/ObserveStoreCreatorProxy.kt
package com.satis.viewmodel.processor

import com.satis.viewmodel.processor.utils.getPackagePath
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import java.lang.Exception
import java.lang.reflect.Constructor
import java.util.LinkedHashMap
import javax.lang.model.element.TypeElement

class ObserveStoreCreatorProxy(private val moduleName: String) {
    private lateinit var typeSpecBuilder: TypeSpec.Builder
    private var getFunSpecBuilder: FunSpec.Builder? = null
    private var propertySpec: PropertySpec.Builder? = null
    private fun initDefaultTypeBuilder() {
        typeSpecBuilder = TypeSpec.classBuilder("ObserveStore_$moduleName")
        val observeStore = ClassName("com.satis.viewmodel.core", "ObserveStore")
        typeSpecBuilder.addSuperinterface(observeStore)
        addProperty()
    }

    private fun addProperty() {
        val classTypeName = Class::class.java.asClassName()
        val lifeOwner = TypeVariableName("out androidx.lifecycle.LifecycleOwner")
        val classParameterizedType = classTypeName.parameterizedBy(lifeOwner)
        val constructorTypeName = Constructor::class.java.asClassName()
        val observerTypeName = TypeVariableName("out com.satis.viewmodel.core.Observer")
        val constructorParameterizedType =
            constructorTypeName.parameterizedBy(observerTypeName)
        val linkedHashMapType = LinkedHashMap::class.java.asClassName().parameterizedBy(
            classParameterizedType,
            constructorParameterizedType
        )

        getFunSpecBuilder = FunSpec.getterBuilder()
            .addStatement("val map = %T()", linkedHashMapType)
            .beginControlFlow("try")


        propertySpec =
            PropertySpec.builder("store", linkedHashMapType, KModifier.OVERRIDE)

    }

    fun put(typeElement: TypeElement) {
        val host = ClassName.bestGuess(typeElement.qualifiedName.toString())
        val className = ClassName(
                    getPackagePath(typeElement),
            typeElement.simpleName.toString() + "_Observe"
        )
        getFunSpecBuilder!!.addStatement("map.put(%T::class.java,%T::class.java.getConstructor())",
            host,
            className)
    }

    fun endMethod() {
        getFunSpecBuilder!!.endControlFlow()
            .beginControlFlow("catch(e:%T)", Exception::class.java)
            .endControlFlow()
            .addStatement("return map")
        propertySpec!!.getter(getFunSpecBuilder!!.build())
    }

    fun build(): TypeSpec {
        typeSpecBuilder.addProperty(propertySpec!!.build())
        return typeSpecBuilder.build()
    }

    init {
        initDefaultTypeBuilder()
    }
}